crates/turborepo-lib/src/run/task_access.rs

```diff
@@ -0,0 +1,244 @@
+use std::{
+    collections::HashMap,
+    fs::{self, File},
+    io::{Error, Write},
+    sync::{Arc, Mutex},
+};
+
+use serde::Deserialize;
+use tracing::{debug, error, warn};
+use turbopath::{AbsoluteSystemPathBuf, PathRelation};
+use turborepo_cache::AsyncCache;
+use turborepo_scm::SCM;
+
+// Environment variable key that will be used to enable, and set the expected
+// trace location
+const TASK_ACCESS_ENV_KEY: &str = "TURBOREPO_TRACE_FILE";
+/// File name where the task is expected to leave a trace result
+const TASK_ACCESS_TRACE_NAME: &str = "trace.json";
+// Path to the config file that will be used to store the trace results
+pub const TASK_ACCESS_CONFIG_PATH: [&str; 2] = [".turbo", "traced-config.json"];
+/// File name where the task is expected to leave a trace result
+const TURBO_CONFIG_FILE: &str = "turbo.json";
+
+use super::ConfigCache;
+use crate::{config::RawTurboJson, unescape::UnescapedString};
+
+#[derive(Debug, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct TaskAccessTraceAccess {
+    pub network: bool,
+    pub file_paths: Vec<UnescapedString>,
+    pub env_var_keys: Vec<UnescapedString>,
+}
+
+#[derive(Debug, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct TaskAccessTraceFile {
+    pub accessed: TaskAccessTraceAccess,
+    pub outputs: Vec<UnescapedString>,
+}
+
+#[derive(Deserialize, Debug)]
+struct PackageJson {
+    scripts: Option<std::collections::HashMap<String, String>>,
+}
+
+pub fn trace_file_path(
+    repo_root: &AbsoluteSystemPathBuf,
+    task_hash: &str,
+) -> AbsoluteSystemPathBuf {
+    repo_root.join_components(&[".turbo", task_hash, TASK_ACCESS_TRACE_NAME])
+}
+
+fn task_access_trace_enabled(repo_root: &AbsoluteSystemPathBuf) -> Result<bool, Error> {
+    // TODO: use the existing config methods here
+    let root_turbo_json_path = &repo_root.join_component(TURBO_CONFIG_FILE);
+    if root_turbo_json_path.exists() {
+        return Ok(false);
+    }
+
+    // read package.json at root
+    let package_json_path = repo_root.join_components(&["package.json"]);
```

Comment:
Is this feature only intended to be used in single package workspaces?

Reply:
For now, yes (but we'll roll this out more broadly soon)